chat-sdk-firebase-adapter/src/main/java/co/chatsdk/firebase/wrappers/ThreadWrapper.java

```diff
@@ -315,37 +315,38 @@ public void messagesOff() {
         FirebaseReferenceManager.shared().removeListeners(ref);
     }
 
-    public Observable<Thread> metaOn () {
-
+    public Observable<Thread> metaOn() {
         return Observable.create((ObservableOnSubscribe<Thread>) e -> {
-            DatabaseReference ref = FirebasePaths.threadMetaRef(model.getEntityID());
-            FirebaseReferenceManager.shared().addRef(ref, ref.addValueEventListener(new FirebaseEventListener().onValue((snapshot, hasValue) -> {
-                if(hasValue &&  snapshot.getValue() instanceof Map) {
-                    Map<String, Object> value = (Map<String, Object>) snapshot.getValue();
-                    for (String key : value.keySet()) {
-                        if (value.get(key) instanceof String) {
-                            model.setMetaValue(key, (String) value.get(key));
-                        }
-                    }
-                }
+
+            metaOff();
+
+            final DatabaseReference threadMetaRef = FirebasePaths.threadMetaRef(model.getEntityID());
+
+            if (FirebaseReferenceManager.shared().isOn(threadMetaRef)) {
                 e.onNext(model);
-            })));
+            }
+
+            ValueEventListener listener = threadMetaRef.addValueEventListener(new FirebaseEventListener().onValue((snapshot, hasValue) -> {
+                if (hasValue && snapshot.getValue() instanceof Map) {
+                    deserializeMeta((Map<String, Object>) snapshot.getValue());
+                    e.onNext(model);
+                }
+            }));
+
+            FirebaseReferenceManager.shared().addRef(threadMetaRef, listener);
+
         }).subscribeOn(Schedulers.single());
     }
 
     public Completable pushMeta() {
         return Completable.create(e -> {
 
-            DatabaseReference ref = FirebasePaths.threadMetaRef(model.getEntityID());
-
-            HashMap<String, String> meta = new HashMap<>();
+            DatabaseReference threadMetaRef = FirebasePaths.threadMetaRef(model.getEntityID());
 
-            List<ThreadMetaValue> values = model.getMetaValues();
-            for(ThreadMetaValue value : values) {
-                meta.put(value.getKey(), value.getValue());
-            }
+            HashMap<String, String> metaMap = new HashMap<>(model.metaMap());
+            Map<String, Object> expandedMetaMap = HashMapHelper.expand(metaMap);
```

Comment:
Why do you do this on two lines? Can't it be:

```
Map<String, Object> expandedMetaMap = HashMapHelper.expand(model.metaMap());
```

FYI you will need to change the expand method to be like this:

```
    public static HashMap<String, Object> expand (Map<String, String> flatMap) {
```